ccp.c

```diff
@@ -38,93 +38,108 @@ void __INLINE__ null_log(struct ccp_datapath *dp, enum ccp_log_level level, cons
     (void)(msg_size);
 }
 
-int ccp_init(struct ccp_datapath *dp) {
+/* 
+ * Return value points to a ccp_datapath with allocated memory. Caller is responsible for 
+ * later freeing this memory with ccp_free. 
+ *
+ * Currently, this return value is a deep copy of the parameter struct with some additional things
+ * Once this works, we should switch back to returning an int and just re-use the struct the caller
+ * passes in so that we can retain return error codes.
+ */
+struct ccp_datapath *ccp_init(struct ccp_datapath *user_dp) {
```

Comment:
While we are changing the api - if we are getting rid of globals, seems like a good opportunity to also get rid of mallocs :)

so this would become
```suggestion
int ccp_init(struct ccp_datapath **ptr, struct ccp_datapath *user_dp);
```

which fixes your error code comment also.

Reply:
if the user is the one allocating the space before they call ccp_init, then why do they need to pass a double pointer?

also, the other part of my comment in the code: is there any specific reason they need to pass in the handlers in a separate structure? 

the user could allocate space for a datapath struct, fill it partially with the handlers, then pass it to ccp_init to fill in the rest of the state etc.

Reply:
hm.. the only thing a bit messy about this is that ccp_datapath contains a pointer to the datapath programs which we malloc. so the user needs to still call ccp_free where we free stuff inside datapath and then call free(datapath) themselves

Reply:
- should be fine to have the caller do the alloc and fill in the function pointers, then have ccp_init fill the rest.
- yes, double pointer not necessary
- ccp_free thing is fine I think, unless you want the user to also alloc the datapath_programs array. I'm kind of in favor of the user alloc'ing, since that removes the last malloc call from libccp. ccp_init would then just check for a valid pointer and the user would have to specify how much memory they have alloc'ed

Reply:
3rd point: i think it would be good if the user either alloc's all of the space or none of it, something in between is a little confusing / tricky. 

the datapath_programs array isn't the last malloc, there' also init_ccp_priv_state for each connection. but now that ccp_active_connections is inside datapath_programs, it's already heap allocated, so we can probably move ccp_priv_state directly into ccp_connection instead of having a pointer to it.

Reply:
something like this?

user code: 
```c
size_t max_connections = 1024;
size_t max_dp_programs = 10;
struct ccp_datapath *datapath = malloc(...);
datapath->ccp_active_connections = malloc(max_connections * sizeof(struct ccp_connection));
datapath->datapath_programs = malloc(max_dp_programs * sizeof(struct DatapathProgram)); 
int ret = ccp_init(datapath, max_connections, max_programs); 
```

libccp:
```c
int ccp_init(struct ccp_datapath *datapath, size_t max_programs, size_t max_connections) {
    ...
    if (!datapath->ccp_active_connections) { ... return  } 
    if (!datapath->datapath_programs) { ... return }
    ...
```

Reply:
it feels a little weird to trust the user to allocate space for ccp_active_connections and datapath_programs correctly, but i don't feel too strongly about it if you think it's fine

Reply:
I think it's fine - the caller of `ccp_init` will most of the time know what they are doing, so if we document properly it will work out. 
I think it's only in userspace with the congestion control api that we needed to be more careful with asking the callers to do things.

Reply:
The one thing I don't like about this is we have no way of checking if they allocated enough memory.. 

One bug I just ran into when adjusting the test to use this new api: I accidentally wrote `malloc( .. * sizeof(struct DatapathProgram *));` (i.e. i was allocating space for pointers not the struct itself).  The error only showed up once we got to install_programs where it tries to memset(0) the datapath programs array. Because not enough space had been allocated, it segfault'd saying we tried to write past the end of the buffer. 

There aren't that many other ways you could screw this up so maybe it's fine but just wanted to bring this up